File: src/stuff/getc.c
#include <stdio.h>
#include <unistd.h>
#include <termios.h>

// http://shtrom.ssji.net/skb/getc.html

int main()
{
    struct termios old_tio, new_tio;
    unsigned char c;

    /* get the terminal settings for stdin */
    tcgetattr(STDIN_FILENO, &old_tio);

    /* we want to keep the old setting to restore them a the end */
    new_tio = old_tio;

    /* disable canonical mode (buffered i/o) and local echo */
    new_tio.c_lflag &= (~ICANON & ~ECHO);
    printf("c_lflag: %ld\n", new_tio.c_lflag);

    /* set the new settings immediately */
    tcsetattr(STDIN_FILENO, TCSANOW, &new_tio);

    do {
         c = getchar();
         printf("*");
         printf("%d ",c);
    } while (c != 'q');

    /* restore the former settings */
    tcsetattr(STDIN_FILENO, TCSANOW, &old_tio);

    return 0;
}

File: src/stuff/getc.nim
import posix, termios

#include <stdio.h>
#include <unistd.h>
#include <termios.h>

# http://shtrom.ssji.net/skb/getc.html

proc getChar(): cint {.importc: "getchar", header: "<stdio.h>".}

proc init() =
  var
    oldtio: Termios
    newtio: Termios
    c: char = cast[char](0x00)

  # get the terminal settings for stdin */
  discard tcGetAttr(STDIN_FILENO, oldtio.addr)

  # we want to keep the old setting to restore them a the end
  newtio = oldtio

  # disable canonical mode (buffered i/o) and local echo
  echo newtio
  newtio.c_lflag = newtio.c_lflag and
                   ((not ICANON.cuint) and (not ECHO.cuint))

  echo newtio

  # set the new settings immediately
  discard tcSetAttr(STDIN_FILENO, TCSANOW, newtio.addr)

  while c != 'q':
    echo "*"
    c = cast[char](getchar())
#    echo c

  # restore the former settings
  discard tcSetAttr(STDIN_FILENO, TCSANOW, oldtio.addr)


init()
